File: java/backend-spring/src/main/java/com/countriesdb/examples/spring/ValidationController07.java
package com.countriesdb.examples.spring;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/validate")
public class ValidationController07 {
    private final CountriesDbClient client;

    public ValidationController07(CountriesDbClient client) {
        this.client = client;
    }

    // 7. Multiple subdivision validation
    @PostMapping("/subdivision/multiple")
    public ResponseEntity<?> validateMultipleSubdivisions(@RequestBody Map<String, Object> payload) {
        String country = (String) payload.get("country");
        if (country == null || country.isBlank()) {
            return ResponseEntity.badRequest().body(Map.of(
                "valid", false,
                "message", "Country code is required"
            ));
        }
        
        @SuppressWarnings("unchecked")
        List<String> subdivisions = (List<String>) payload.get("subdivisions");
        if (subdivisions == null || subdivisions.isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of(
                "valid", false,
                "message", "Subdivisions list is required"
            ));
        }
        
        Map<String, Object> result = client.validateSubdivision(Map.of(
            "country", country,
            "code", subdivisions
        ));
        
        @SuppressWarnings("unchecked")
        List<Map<String, Object>> results = (List<Map<String, Object>>) result.get("results");
        if (results != null) {
            for (Map<String, Object> item : results) {
                if (!Boolean.TRUE.equals(item.get("valid"))) {
                    return ResponseEntity.unprocessableEntity().body(result);
                }
            }
        }
        
        return ResponseEntity.ok(Map.of(
            "valid", true,
            "message", "All subdivisions validated",
            "payload", payload
        ));
    }
}

File: csharp/backend-aspnetcore/Controllers/ValidationController07.cs
using Microsoft.AspNetCore.Mvc;
using System.Text.Json;

namespace BackendAspNetCore.Controllers;

[ApiController]
[Route("api/validate")]
public class ValidationController07 : ControllerBase
{
    private readonly CountriesDbClient _client;

    public ValidationController07(CountriesDbClient client)
    {
        _client = client;
    }

    // 7. Multiple subdivision validation
    [HttpPost("subdivision/multiple")]
    public async Task<IActionResult> ValidateMultipleSubdivisions([FromBody] Dictionary<string, object> payload)
    {
        var country = payload.GetValueOrDefault("country")?.ToString();
        if (string.IsNullOrWhiteSpace(country))
        {
            return BadRequest(new Dictionary<string, object>
            {
                ["valid"] = false,
                ["message"] = "Country code is required"
            });
        }

        if (!payload.ContainsKey("subdivisions") || payload["subdivisions"] is not JsonElement subdivisionsElement)
        {
            return BadRequest(new Dictionary<string, object>
            {
                ["valid"] = false,
                ["message"] = "Subdivisions array is required"
            });
        }

        var subdivisions = JsonSerializer.Deserialize<string[]>(subdivisionsElement.GetRawText());
        if (subdivisions == null || subdivisions.Length == 0)
        {
            return BadRequest(new Dictionary<string, object>
            {
                ["valid"] = false,
                ["message"] = "At least one subdivision code is required"
            });
        }

        var result = await _client.ValidateSubdivisionsAsync(new Dictionary<string, object>
        {
            ["country"] = country,
            ["code"] = subdivisions
        });

        if (result.Results == null)
        {
            return UnprocessableEntity(new ValidationResponse { Valid = false, Message = "Validation failed" });
        }

        foreach (var item in result.Results)
        {
            if (!item.Valid)
            {
                return UnprocessableEntity(new
                {
                    valid = false,
                    message = item.Message ?? "Invalid subdivision code",
                    results = result.Results
                });
            }
        }

        return Ok(new Dictionary<string, object>
        {
            ["valid"] = true,
            ["message"] = "All subdivisions validated",
            ["payload"] = payload
        });
    }
}

